Add option to reverse FlowContainer fill direction (HFlow bottom-to-top, VFlow right-to-left)

TheSecondReal0 opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->
Adds an option to make horizontal FlowContainers add lines starting from the bottom instead of the top or vertical flow containers add lines starting from the right instead of left. 
Video:

https://user-images.githubusercontent.com/66881186/222279524-b493f57e-2a49-45ad-b5f2-9cfcaae50ff3.mp4

- Implements and closes https://github.com/godotengine/godot-proposals/issues/6386

Calinou commented:
@TheSecondReal0 Remember that for GitHub video previews to work, you need to have a blank line before *and* after the video URL. I edited your post accordingly, but remember to do this in the future :slightly_smiling_face:

KoBeWi commented:
It's rather odd to have it only in HFlowContainer. The property could be renamed to *e.g.* `reverse_fill` and implemented in VFlow as right-to-left.

YuriSizov commented:
Layout direction is an internationalization option, you shouldn't mix the meanings here. It's effect should be applied on top of whatever the behavior is according to other options.

TheSecondReal0 commented:
I agree layout direction shouldn't be used in this way, just acknowledging the workaround. I'll update this PR to handle vertical flows + rename variable

TheSecondReal0 commented:
Renamed option to `reverse_fill` and made it affect vertical flow containers as well, not sure about the best way to describe it in documentation. When used in vertical containers it technically fills columns in the direction opposite `layout_direction` but that's less readable than "right to left". For now I just added another sentence but there's probably a better way to word it.

I added this edge case to be consistent with `HFlowContainer` switching directions when `layout_direction` changes, which doesn't seem to be explicitly documented anywhere.

AThousandShips commented:
You will need to remove the commit messages from the squash from your commit, all th ones copied from the merge, the message is much too big

TheSecondReal0 committed:
Add option to reverse FlowContainer fill direction (HFlow bottom-to-top, VFlow right-to-left)

AThousandShips commented:
The new message should probably be cleaned up as well, as it mostly contains minor progress descriptions not relevant to the commit like "ran --doctool"

AThousandShips commented on doc/classes/FlowContainer.xml:
```diff
@@ -21,6 +21,10 @@
 			The alignment of the container's children (must be one of [constant ALIGNMENT_BEGIN], [constant ALIGNMENT_CENTER], or [constant ALIGNMENT_END]).
 		</member>
+		<member name="reverse_fill" type="bool" setter="set_reverse_fill" getter="is_reverse_fill" default="false">
+			If [code]true[/code], reverses fill direction. Horizontal [FlowContainer] will fill rows bottom to top, vertical [FlowContainer] will fill columns right to left.
```
```suggestion
			If [code]true[/code], reverses fill direction. Horizontal [FlowContainer]s will fill rows bottom to top, vertical [FlowContainer]s will fill columns right to left.
```
Incomplete sentence

TheSecondReal0 replied:
Fixed

KoBeWi commented on scene/gui/flow_container.cpp:
```diff
@@ -327,6 +330,16 @@ bool FlowContainer::is_vertical() const {
 
+void FlowContainer::set_reverse_fill(bool p_reverse_fill) {
+	reverse_fill = p_reverse_fill;
+	update_minimum_size();
```
This property doesn't affect minimum size, no?

KoBeWi commented on scene/gui/flow_container.cpp:
```diff
@@ -327,6 +330,16 @@ bool FlowContainer::is_vertical() const {
 }
 
+void FlowContainer::set_reverse_fill(bool p_reverse_fill) {
+	reverse_fill = p_reverse_fill;
```
Should check if the value is different (see `set_alignment()`).

TheSecondReal0 commented:
Thanks for the review!

Geometror commented on scene/gui/flow_container.cpp:
```diff
@@ -338,11 +353,14 @@ void FlowContainer::_bind_methods() {
 	BIND_ENUM_CONSTANT(ALIGNMENT_END);
 
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "alignment", PROPERTY_HINT_ENUM, "Begin,Center,End"), "set_alignment", "get_alignment");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "reverse_fill"), "set_reverse_fill", "is_reverse_fill");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "vertical"), "set_vertical", "is_vertical");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "vertical"), "set_vertical", "is_vertical");
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "reverse_fill"), "set_reverse_fill", "is_reverse_fill");
```
Small nitpick: I would put this at the end.

Geometror reviewed (approved):
Tested for all combinations of alignment, layout direction and reverse fill. Nice work!

suddjian commented:
Just want to say how thrilled I am to be searching for my use case and find a PR adding the feature that was just approved 7 hours ago.

AThousandShips commented:
Needs a rebse to fix conflicts

TheSecondReal0 commented:
Addressed @Geometror's nitpick (totally agree about ordering) and rebased. Hopefully ready to merge 🙏

AThousandShips commented on doc/classes/FlowContainer.xml:
```diff
@@ -21,6 +21,10 @@
 		</member>
+		<member name="reverse_fill" type="bool" setter="set_reverse_fill" getter="is_reverse_fill" default="false">
+			If [code]true[/code], reverses fill direction. Horizontal [FlowContainer]s will fill rows bottom to top, vertical [FlowContainer]s will fill columns right to left.
+			When using a vertical [FlowContainer] with a right to left [code]layout_direction[/code], columns will fill left to right instead.
```
```suggestion
			When using a vertical [FlowContainer] with a right to left [member Control.layout_direction], columns will fill left to right instead.
```

TheSecondReal0 replied:
Done

akien-mga commented:
Thanks!
